### PR TITLE
Fix VehicleWheel sinks into other colliders

### LINKED ISSUE #45339: VehicleWheels sinks into other colliders
**Godot version:**
3.2.3-stable

**OS/device including version:**
Linux

**Issue description:**
Anytime I use VehicleWheel, it sinks into the ground as seen here:
image

The wheel is made to fit she mesh like so:
image

**Steps to reproduce:**
1. Start Godot's truck town demo
2. Select the truck to drive
3. Observe it sink into the ground

**Minimal reproduction project:**
Truck town has it all over:
https://github.com/godotengine/godot-demo-projects/tree/master/3d/truck_town

### PR DESCRIPTION
**Pull request by MrMinimal**
Fixes #45339.

Tested in [truck town](https://github.com/godotengine/godot-demo-projects/tree/master/3d/truck_town) and my simple 
[demo.zip](https://github.com/godotengine/godot/files/6217491/demo.zip).

### CODE REVIEW

**Comment by MrMinimal** [comment]
I couldn't run master branch because of too many exceptions, I hope it's cool that I used the 3.x branch.

The change is so small, a cherry pick for master could even work.

**Comment by MrMinimal** [comment]
I have updated the PR now to merge into master.

@Reviewer
If I open the editor (`-e` flag) on the master branch I get exceptions, so I was only able to test this PR when running the demo.zip directly (`--path` flag).

File: scene/3d/vehicle_body_3d.cpp [hunk 1]
```diff
@@ -399,7 +399,6 @@ real_t VehicleBody3D::_ray_cast(int p_idx, PhysicsDirectBodyState3D *s) {
 	Vector3 source = wheel.m_raycastInfo.m_hardPointWS;
 	wheel.m_raycastInfo.m_contactPointWS = source + rayvector;
 	const Vector3 &target = wheel.m_raycastInfo.m_contactPointWS;
-	source -= wheel.m_wheelRadius * wheel.m_raycastInfo.m_wheelDirectionWS;
```
Comment (pouleyKetchoupp) [inline scene/3d/vehicle_body_3d.cpp:402]: Why is this line removed? (genuine question, it looks like it's the main part of the fix and I don't know this code enough to evaluate the change).

It would generally help to describe a bit more what the fix is in the PR description and the commit message to help for review and future contributions.

Comment (VikramSGIT) [reply scene/3d/vehicle_body_3d.cpp:402]: You are right Sir :+1: , that was the main part of the fix. To know more information about this fix and to know more about VehicleBody you can visit my 
GSoC proposal :
https://docs.google.com/document/d/1P6wNMr-bsVEsX_Eh4-SKHb_rfNGzdujDs-0ldeD_45Q/edit#heading=h.g5s3esky2w5f. 
Can also do a review!!
Tnx :smiling_face_with_three_hearts: !!

**Comment by b7g** [comment]
@VikramSGIT Will you be fixing this?

**Comment by VikramSGIT** [comment]
I guess it has been fixed, If not I can. Thank you

**Comment by MrMinimal** [comment]
@VikramSGIT I'd appreciate your input here, I can't answer @pouleyKetchoupp question because it was mostly trial and error.

You seem more knowledgeable, would you mind explaining it or fixing it properly?

**Comment by AgentSmith0** [comment]
Any updates to this?

**Comment by Calinou** [comment]
This PR may actually not be needed if following the steps outlined here resolves the issue: https://github.com/godotengine/godot/issues/45339#issuecomment-1050273752

@MrMinimal Can you test this on your end? The Truck Town demo hasn't been ported to `master` yet, but I'd like someone to make sure the above steps actually work as expected.

**Comment by MrMinimal** [comment]
@Calinou Can't reproduce the fix mentioned. I think what happens on their end is the stiffness is high enough for the suspension to not be compressed. The sinking is only visible when compression happens though. IMHO the PR is still valid.

**Review by Calinou — COMMENTED** [review]
Tested locally in https://github.com/godotengine/godot-demo-projects/tree/master/3d/truck_town (rebased on top of `master` https://github.com/godotengine/godot/commit/7d2ca2d8ac49cde9767e00b70f9eaf1920eb266d), it **sometimes** works as expected.

Sometimes, your wheels can completely sink into the ground if you use the tow truck, run into a building then reverse:

image

This doesn't occur with the `master` branch, even though wheels will be constantly slightly sinking into the ground.

Before | After
-|-
Screenshot_20240229_234954 webp | Screenshot_20240301_000546 webp

Before | After
-|-
Screenshot_20240229_235010 webp | Screenshot_20240301_000556 webp

Before | After
-|-
Screenshot_20240229_235028 webp | Screenshot_20240301_000615 webp

In short, this PR improves upon the "general" case but worsens the "special" case when your vehicle collides with something.

My rebased branch for reference: https://github.com/Calinou/godot/tree/fix-wheels-sinking

**Comment by fire** [comment]
@aaronfranke, since you were looking into vehicles, do you have any particular thoughts?

**Comment by aaronfranke** [comment]
@fire I do not. I've only been looking into how to represent the data of vehicles including wheels in an interoperable way, not the precise implementation details of how wheels work.

**Comment by skorpnok** [comment]
So from reading the code and the proposed fix a little bit the problem stems from a discrepancy between `raylen` in line 396 and `param` in line 415. But in line 402 (which this proposal removes) it adds an offset to the begin of the raycast to make it harder for the wheel to fall through the ground. But this offset is not accounted for when `param` is computed/used.

So replacing the line
`param = source.distance_to(rr.position) / source.distance_to(target);`
with
`param = (source.distance_to(rr.position) - wheel.m_wheelRadius) / raylen;`
without any of the other changes proposed here fixes this issue. Tested locally.

The current proposal gets rid of the discrepancy but also gets rid of the safety margin as observed by @Calinou

I can make a new PR for the change if that's easier.
